### DISCUSSION TITLE
MQTT with Teensy4.1

### DISCUSSION THREAD

**Discussion by Mike0675** [cesanta/mongoose#2491] [Ask questions, get answers] [answered]
Hi,  I´m happy to see that Teensy is now supported by mongoose.

Http-server is workin well, but I´m incapable to get MQTT work with Raspbian´s Mosquitto.

Once Mosquitto accept Data from Teensy, then again nothing happens.

Im adapting the new Arduino-Example + Adding MQTT 

`
// Run MQTT:
int topts = MG_TIMER_REPEAT | MG_TIMER_RUN_NOW;
mg_timer_add(&mgr, 3000, topts, mqtt_timer_fn, &mgr);

/////////////////////////////////////////////////////////////////////////////////////////////////

static void mqtt_timer_fn(void *arg)
{	
	struct mg_mgr *mgr = (struct mg_mgr *)arg;
	struct mg_mqtt_opts opts;
	memset(&opts, 0, sizeof(opts));

	opts.user = mg_str(mqtt_user);
	opts.pass = mg_str(mqtt_pass);
	opts.client_id = mg_str("jlkdsfjlksdjfölakjs");
	opts.topic = mg_str(HOSTNAME);
	opts.message = mg_str("jlkdsfjlksdjfölakjs");
	opts.qos = s_qos;
	opts.version = 4;
	opts.keepalive = 1000;
	opts.retain = false;
	opts.clean = true;

	if (mif.state == MG_TCPIP_STATE_READY && mqtt_connection == NULL)
	{
		MG_INFO(("Starting MQTT connection..."));
		MG_INFO(("mqtt_id :"));

		char *mqtt_id = "Teensy41_";
		char RandomNumber[10];
		randomSeed(micros());
		itoa(random(10000), RandomNumber, DEC);
		strcat(mqtt_id, RandomNumber);

		MG_INFO((mqtt_id));

		mqtt_connection = mg_mqtt_connect(mgr, mqtt_url, &opts, mqtt_event_handler, NULL);
	}
}

///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

void mqtt_event_handler(struct mg_connection *c_mqtt, int ev, void *ev_data, void *fn_data)
{

	if (ev == MG_EV_OPEN)
	{
		MG_INFO(("%lu CREATED", c_mqtt->id));
		// c->is_hexdumping = 1;
	}
	else if (ev == MG_EV_ERROR)
	{
		// On error, log error message
		MG_ERROR(("%lu ERROR %s", c_mqtt->id, (char *)ev_data));
	}
	else if (ev == MG_EV_CONNECT)
	{
	}
	else if (ev == MG_EV_MQTT_OPEN)
	{
		// MQTT connect is successful
		MG_INFO(("%lu CONNECTED to %s", c_mqtt->id, mqtt_url));
		
		opts.topic = mg_str("Test/Test1/Test2");
		opts.message = mg_str("test");
		mg_mqtt_pub(c_mqtt, &opts);
	}

	else if (ev == MG_EV_MQTT_MSG)
	{
		// When we get echo response, print it
		struct mg_mqtt_message *mm = (struct mg_mqtt_message *)ev_data;
		MG_INFO(("%lu RECEIVED %.*s <- %.*s", c_mqtt->id, (int)mm->data.len,
				 mm->data.ptr, (int)mm->topic.len, mm->topic.ptr));

		
	}
	else if (ev == MG_EV_CLOSE)
	{
		MG_INFO(("%lu CLOSED", c_mqtt->id));
		c_mqtt = NULL; // Mark that we're closed
	}
	(void)fn_data;
}
`

**Comment by scaprile** [comment, marked as answer]
I don't know what "Once Mosquitto accept Data from Teensy" nor ""nothing happens" mean.
I suggest you sniff your network (Wireshark) and see if what you think it is, actually is. Did you happen to try to connect to a public broker that everybody knows is correctly configured and working fine, like HiveMQ, as we use in our [examples](https://github.com/cesanta/mongoose/tree/master/examples) and [tutorials](https://mongoose.ws/tutorials/) ?
Our MQTT examples are tested and do not depend on Teensy nor any other hardware.
If, after your network inspection, you find an actual issue, feel free to open an issue, fill the issue template, and provide us with pertinent information (including useful network captures) in order to reproduce and fix the issue. We don't inspect your code for errors, please send the minimum piece of code that can reproduce the issue
